02-peer-protocol.md

```diff
@@ -520,61 +527,113 @@ Fulfilling an HTLC is different: knowledge of the preimage is by definition irre
- A if the payment does not complete.
-
-For example, node A might offer node B an HTLC with a timeout of 3 days, and node B might
-offer node C the same HTLC with a timeout of 2 days:
-
-```
-    3 days timeout        2 days timeout
-A ------------------> B ------------------> C 
-```
-
-The difference in timeouts is called `cltv_expiry_delta` in 
-[BOLT #7](07-routing-gossip.md).
-
-This difference is important: after 2 days B can try to
-remove the offer to C even if C is unresponsive, by broadcasting the
-commitment transaction it has with C and spending the HTLC output.
-Even though C might race to try to use its payment preimage at that point to
-also spend the HTLC, it should be resolved well before the 3 day
-deadline so B can either redeem the HTLC off A or close it.
+Consider the following scenario, where A sends an HTLC to B, who
+forwards to C, who delivers the goods as soon as the payment is
+received.
+
+1.  C needs to be sure that the HTLC from B cannot time out, even if B becomes
+    unresponsive; i.e. C can fulfill the incoming HTLC on-chain before B can
+    time it out on-chain.
+
+2.  B needs to be sure that if C fulfills the HTLC from B, it can fulfill the
+    HTLC from A.  i.e. B can get the preimage from C and fulfill incoming the
+    HTLC on-chain before B can time it out on-chain.
+
+The critical setting here is the `cltv_expiry_delta` (in both
+[`open_channel`](#the-open_channel-message) and in
+[BOLT #7](07-routing-gossip.md#the-channel_update-message)): it's used both
+for the minimum difference in HTLC CLTV timeouts in the forwarding case (B)
+and the minimum difference between HTLC CLTV timeout and the current block
+height for the terminal case (C).
+
+Note that if this value is too low for a channel, the risk is only to the node
+*accepting* the HTLC, not the node offering it.  Most other channel properties
+(eg. fees) are set by the node offering the HTLC; for this reason, nodes
+exchange their `cltv_expiry_delta` on channel creation and their
+`channel_update` will contain the other node's `cltv_expiry_delta`
+in [BOLT #7](07-routing-gossip.md#the-channel_update-message).
+
+We can derive the worst-case number of blocks between outgoing and
+incoming HTLC resolution, given a few assumptions:
+
+* A worst-case reorganization depth `R` blocks
+* A grace-period `G` blocks after HTLC timeout before we give up on
+  the peer and drop to chain.
+* A number of blocks `S` between transaction broadcast and the
+  transaction being included in a block.
+
+The worst case is for a forwarding node (B) which takes the longest
+possible time to spot the outgoing HTLC fulfillment, and then takes
+the longest possible time to redeem it on-chain:
+
+1. The B->C HTLC times out at block `N`, and B waits `G` blocks until
+   it gives up waiting for C.  B or C commits to the blockchain,
+   and B spends HTLC.
+2. Bad case: C wins the race and fulfills the HTLC, B only sees
```

Comment:
What if neither B nor C's block makes it into the first block after `G`? Should this take into account `S`, B's time to confirm? (or C's in the case that C wins the race at the latest possible block)

Reply:
Conservatively, I think you're right.  B will spend via time out (worst case) at N+G+S+R.  So C can fulfill at any time <= N+G+S+R.

That gives us 3R+2G+2S; that changes our recommendation from 9 to 12.

However, if @Roasbeef wants a higher S value, then it increases again (as would our recommended final cltv value).